connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java

```diff
@@ -758,9 +758,11 @@ public void run() {
     }
 
     // Common handling for requests that get config data. Checks if we are in sync with the current config, which allows
-    // us to answer requests directly. If we are not, handles invoking the callback with the appropriate error.
+    // us to answer requests directly. If we are not, handles invoking the callback with the appropriate error;
+    // Note that the configState's offset can be bigger than assignment's offset; Because if updating an existing
+    // connector, the connector can pick up the config change without a rebalance;
     private boolean checkConfigSynced(Callback<?> callback) {
-        if (assignment == null || configState.offset() != assignment.offset()) {
+        if (assignment == null || configState.offset() < assignment.offset()) {
```

Comment:
I don't think this check will be correct either. For example, what if the `configState` was updated, a rebalance runs and results in the `configState.offset()` and `assignment.offset()` being synced, and then a whole new set of `configState` data is pulled in without a rebalance being performed (one may be requested, but if the rebalance and assignment process doesn't complete, the test here would pass, but the assignment would be out of date).

The existing check should already validate that `configState.offset() >= assignment.offset()` anyway, right? Shouldn't we be unable to join a group and get an assignment with the specified offset unless we have the configs up to that offset? Or, I guess we can get one if the assignment has failed, but then we could just check `assignment.error()`?

I think if you wanted to get read-after-write without any possibility of error codes and avoid rebalances, we might need to track offsets in a more complicated way (e.g., track what offsets we used when re-joining that generated the assignment vs what is the latest offset we've read up to).